File: server/keeper_action.go
package main

import "math"

// active_goalie_no_ball_ai, Amiga 0xfcf4–0xfd60: immediate actions.
func (s *State) keeperAction(i, distance, random int) (*interaction, float64, float64) {
	const unit = 22.4 / 576
	p, b := &s.Players[i], &s.Ball
	if b.Owner >= 0 && s.Players[b.Owner].Team == p.Team {
		return nil, 0, 0
	}
	qx, qz, vx, vz := b.X, b.Z, b.VX, b.VZ
	if b.Owner >= 0 {
		q := s.Players[b.Owner]
		qx, qz, vx, vz = q.X, q.Z, q.moveX, q.moveZ
	}
	side := p.Team
	if s.Period == 2 {
		side ^= 1
	}
	ymin, ymax := 960, 1104
	if side != 0 {
		ymin, ymax = 48, 192
	}
	x, y := int(math.Round(qz/unit+320)), int(math.Round(576-qx/unit))
	attack := s.worldInViewport(p.X, p.Z, 0) && distance <= 8*actionSustain(p.Stats[3])
	if attack {
		shift := max(0, min(2, (p.Stats[7]-100)/50))
		dx, dy := int(math.Round(vz/unit/25)), -int(math.Round(vx/unit/25))
		for shift > 0 {
			next := y + dy*(1<<shift)
			if (side != 0 || next <= ymax) && next >= ymin {
				break
			}
			shift--
		}
		x += dx * (1 << shift)
		y += dy * (1 << shift)
	} else {
		if vx != 0 || vz != 0 || p.Stats[0]/2 <= random {
			return nil, 0, 0
		}
		x = max(160, min(480, x))
		y = max(ymin, min(ymax, y))
	}
	tx, tz := float64(576-y)*unit, float64(x-320)*unit
	dx, dz := int(math.Round(tx/unit)-math.Round(p.X/unit)), int(math.Round(tz/unit)-math.Round(p.Z/unit))
	a := &interaction{attack: attack}
	if absInt(dx) > absInt(dz)/2 {
		a.x = math.Copysign(1, float64(dx))
	}
	if absInt(dz) > absInt(dx)/2 {
		a.z = math.Copysign(1, float64(dz))
	}
	return a, tx, tz
}

File: src/keeper-action.js
import { actionSustain } from "./attributes.js";
import { worldInViewport } from "./visibility.js";
const unit = 22.4 / 576;

// active_goalie_no_ball_ai, Amiga 0xfcf4–0xfd60: immediate actions.
export function keeperAction(s, i, distance, random) {
  const p = s.players[i],
    b = s.ball;
  if (b.owner >= 0 && s.players[b.owner].team === p.team) return null;
  const q = b.owner < 0 ? b : s.players[b.owner];
  const vx = b.owner < 0 ? b.vx : q.moveX || 0,
    vz = b.owner < 0 ? b.vz : q.moveZ || 0;
  const side = p.team ^ (s.period === 2 ? 1 : 0);
  const ymin = side === 0 ? 960 : 48,
    ymax = side === 0 ? 1104 : 192;
  let x = Math.round(q.z / unit + 320),
    y = Math.round(576 - q.x / unit);
  const attack =
    worldInViewport(s, p) && distance <= 8 * actionSustain(p.stats[3]);
  if (attack) {
    let shift = Math.max(0, Math.min(2, Math.floor((p.stats[7] - 100) / 50)));
    const dx = Math.round(vz / unit / 25),
      dy = -Math.round(vx / unit / 25);
    while (shift > 0) {
      const next = y + dy * 2 ** shift;
      if ((side !== 0 || next <= ymax) && next >= ymin) break;
      shift--;
    }
    x += dx * 2 ** shift;
    y += dy * 2 ** shift;
  } else {
    if (vx || vz || Math.floor(p.stats[0] / 2) <= random) return null;
    x = Math.max(160, Math.min(480, x));
    y = Math.max(ymin, Math.min(ymax, y));
  }
  const tx = (576 - y) * unit,
    tz = (x - 320) * unit;
  const dx = Math.round(tx / unit) - Math.round(p.x / unit),
    dz = Math.round(tz / unit) - Math.round(p.z / unit);
  return {
    attack,
    tx,
    tz,
    x: Math.abs(dx) > Math.floor(Math.abs(dz) / 2) ? Math.sign(dx) : 0,
    z: Math.abs(dz) > Math.floor(Math.abs(dx) / 2) ? Math.sign(dz) : 0,
  };
}
